### PR TITLE
Fix issue 15653 - IFTI fails for immutable parameter

### PR DESCRIPTION
**Pull request by JinShil**
Revival of #5452

Special thanks to @9rnsr

### CODE REVIEW

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -4324,6 +4324,14 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                 }
             }
 
+            bool isTopRef(Type t)
+            {
+                auto tb = t.baseElemOf();
+                return tb.ty == Tclass ||
+                       tb.ty == Taarray ||
+                       tb.ty == Tstruct && tb.hasPointers();
```
Comment (ibuclaw) [inline src/dmd/dtemplate.d]: Do we need fail tests for these types?

File: src/dmd/dtemplate.d [hunk 2]
```diff
@@ -4324,6 +4324,14 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                 }
             }
 
+            bool isTopRef(Type t)
```
Comment (ibuclaw) [inline src/dmd/dtemplate.d]: Maybe just need a small comment on what this function returns.  True if the `t` is a reference type, or an array of reference types.

**Review by ibuclaw — CHANGES_REQUESTED** [review]
As per comments

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/dtemplate.d]: The failure conditions are covered using the `compiles` traits (e.g. `static assert(!__traits(compiles, bar!U(u)), U.stringof);`), in the runnable test.  Is that not sufficient?

Comment (ibuclaw) [reply src/dmd/dtemplate.d]: On second look, ok, there are three types.

- `Object` covers `tb.ty == Tclass`
- `Y` covers `tb.ty == Tstruct && tb.hasPointers()`
- `Y[2]` covers `t.baseElemOf()`

So only missing a check for AAs, adding `int[int]` should be enough by the looks.

Comment (JinShil) [reply src/dmd/dtemplate.d]: Done.  Thank you!

**Review by ibuclaw — APPROVED** [review]

**Review by wilzbach — APPROVED** [review]

**Pull request merged**